index.d.ts

```diff
@@ -0,0 +1,476 @@
+// Type definitions for Node OAuth2 Server 4.0
+// Definitions by:  Robbie Van Gorkom <https://github.com/vangorra>,
+//                  Charles Irick <https://github.com/cirick>,
+//                  Daniel Fischer <https://github.com/d-fischer>,
+//                  Vitor Santos <https://github.com/rvitorsantos>
+
+import * as http from 'http';
+
+/**
+ * Represents an OAuth2 server instance.
+ */
+declare class OAuth2Server {
+    static OAuth2Server: typeof OAuth2Server;
+
+    /**
+     * Instantiates OAuth2Server using the supplied model
+     */
+    constructor(options: OAuth2Server.ServerOptions);
+
+    /**
+     * Authenticates a request.
+     */
+    authenticate(
+        request: OAuth2Server.Request,
+        response: OAuth2Server.Response,
+        options?: OAuth2Server.AuthenticateOptions,
+        callback?: OAuth2Server.Callback<OAuth2Server.Token>
+    ): Promise<OAuth2Server.Token>;
+
+    /**
+     * Authorizes a token request.
+     */
+    authorize(
+        request: OAuth2Server.Request,
+        response: OAuth2Server.Response,
+        options?: OAuth2Server.AuthorizeOptions,
+        callback?: OAuth2Server.Callback<OAuth2Server.AuthorizationCode>
+    ): Promise<OAuth2Server.AuthorizationCode>;
+
+    /**
+     * Retrieves a new token for an authorized token request.
+     */
+    token(
+        request: OAuth2Server.Request,
+        response: OAuth2Server.Response,
+        options?: OAuth2Server.TokenOptions,
+        callback?: OAuth2Server.Callback<OAuth2Server.Token>
+    ): Promise<OAuth2Server.Token>;
+}
+
+declare namespace OAuth2Server {
+    /**
+     * Represents an incoming HTTP request.
+     */
+    class Request {
+        body?: any;
+        headers?: { [key: string]: string; } | undefined;
+        method?: string | undefined;
+        query?: { [key: string]: string; } | undefined;
+
+        /**
+         * Instantiates Request using the supplied options.
+         *
+         */
+        constructor(options?: { [key: string]: any } | http.ClientRequest);
```

Comment:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9b9cd671114a2a5178809798d8e7f4d8ca6c2671/types/express-serve-static-core/index.d.ts#L375

```suggestion
        constructor(options?: { [key: string]: any } | http.IncomingMessage);
```

Reply:
Good catch. Thanks!